rfcs/RFC-Uploads.md

```diff
@@ -0,0 +1,401 @@
+# User Uploads
+
+## For the cellxgene Data Portal
+
+**Authors:** [Trent Smith](mailto:trent.smith@chanzuckerberg.com)
+
+**Approvers:** [Arathi Mani](mailto:arathi.mani@chanzuckerberg.com), [Timmy Huang](mailto:thuang@chanzuckerberg.com),
+[Marcus Kinsella](mailto:mkinsella@chanzuckerberg.com), [Eduardo Lopez](mailto:elopez@chanzuckerberg.com), [Brian Raymore](mailto:braymor@chanzuckerberg.com)
+[Sara Gerber](mailto:sara.gerber@chanzuckerberg.com)
+
+## tl;dr
+
+A user can upload files of any size to Data Portal using either a shared link from their cloud provider or directly from their browser.
+
+## Glossary
+
+- CSP(Cloud Storage Provider) - Cloud storage services such as Dropbox or Google Drive.
+- Shared Link - A link from a CSP that gives the bearer of that link access to private data.
+- DP(Data Portal) - The cellxgene Data Portal.
+- Small File - a file < 5 GB in size.
+- Large File - a file > 5 GB in size.
+
+## Problem Statement | Background
+
+One of the primary goals of the DP is to allow users to upload their own files. These files consist of data sets as well
+as attestation files. These data sets can vary in size from a few MB to several GB. Users may want to upload from their local machine or their CSP. For users that want to upload from their local machine, they must be able to resume an
+upload in the event their connection is interrupted. For users that want to upload files from their CSP, they can provide a shared link from their CSP that is used to upload directly to S3.
+
+This specification does not cover the validation of the data or the process for a user to complete a submission.
+
+## Product Requirements
+
+1. A user can upload a file to DP using a shared link from their Dropbox.
+1. A user can upload a file to DP using a shared link from their Google Drive.
+1. A user can upload a file from their local machine using the DP browser.
+1. A user can resume an upload without completely restarting.
```

Comment:
I think the resumable feature needs some more description in this document.
Here are some of the issues/questions  that come to mind.

If there are 100 parts to the multi-part upload, and the user has uploaded 40 of them before their laptop runs out of battery and dies.  When they get their laptop up and running again, and want to resume that transfer, how will that happen?   What API will be called?  Will the status still be "Uploading" or will it have moved to "Error".  Perhaps a new stage like "Paused"?

It seems like the client will need access to the multi part UploadId, so it can be resumed.   The client will also need to know which parts have been uploaded so that it can know not to resend those.

Reply:
AWS multipart upload handles all of this. https://docs.aws.amazon.com/AmazonS3/latest/dev/mpuoverview.html, the S3 API supplies all of the endpoint we need to start, stop, resume, and abort a multipart upload. The exact order of operation, I didn't dive deep into it. All we need to do is give the client sufficient privilege to the S3 API to use it.

Reply:
From the link you shared:  "When you send a request to initiate a multipart upload, Amazon S3 returns a response with an upload ID, which is a unique identifier for your multipart upload. You must include this upload ID whenever you upload parts, list the parts, complete an upload, or stop an upload."   My question was how to manage the upload ID.  The client gets it when it initiates the multi part upload.  If we don't do anything,  and if the client crashes, then it's lost, and the upload can't be resumed.  At least this is how I understand it.

Reply:
That was my perception as well - but I'm still parsing whether it could be acquired from 
[ListMultipartUploads](https://docs.aws.amazon.com/AmazonS3/latest/API/API_ListMultipartUploads.html) described as:

> This operation lists in-progress multipart uploads. An in-progress multipart upload is a multipart upload that has been initiated using the Initiate Multipart Upload request, but has not yet been completed or aborted.